modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextFieldSkin.java

```diff
@@ -802,25 +802,41 @@ private void updateTextPos() {
                 // empty), then we align the Text node so that the caret will
                 // appear at the left of the centered prompt.
                 newX = midPoint - promptNode.getLayoutBounds().getWidth() / 2;
-                promptNode.setLayoutX(newX);
+                if (newX > 0) {
+                    promptNode.setLayoutX(newX);
+                } else {
+                    // Align to left when prompt text length is more than text field width
+                    promptNode.setLayoutX(caretWidth / 2);
+                }
             } else {
                 newX = midPoint - textNodeWidth / 2;
-            }
-            // Update if there is space on the right
-            if (newX + textNodeWidth <= textRight.get()) {
-                textTranslateX.set(newX);
+                // Update if there is space on the right
+                if (newX + textNodeWidth <= textRight.get()) {
+                    textTranslateX.set(newX);
+                } else if (newX < 0 && oldX > 1) {
+                    textTranslateX.set(caretWidth / 2);
+                }
```

Comment:
I get the impression this code also needs to use `caretWidth / 2` instead of `0` and `1`.

ie:

```suggestion
                // Update if there is space on the right
                if (newX + textNodeWidth <= textRight.get() - caretWidth / 2) {
                    textTranslateX.set(newX);
                } else if (newX < caretWidth / 2 && oldX > caretWidth / 2) {
                    textTranslateX.set(caretWidth / 2);
                }
```

It would only work for very narrow caret's otherwise, and for wide caret's it would have some unexpected positioning in the edge case where text almost fits.

Reply:
> I get the impression this code also needs to use `caretWidth / 2` instead of `0` and `1`.
> 
Here also same comment as above for the `newX < caretWidth/2` condition. But while comparing the `oldX`, `caretWidth/2` should be considered instead of 1 and also `caretWidth / 2` is subtracted from `textRight` value in the previous if condition as you suggested. These updates have been done in the code.

> I would also recommend testing the code on Windows with the screen scale set to 225%, as it might show issues related to fractional scale.

I will test this latest code updates.

Reply:
Tested in windows with screen scale set to as high as 225%. Couldn't reproduce any issues with updated code.

Reply:
thank you for checking!